### PR TITLE
Removing Avatar from CopyFrom in Account

### PR DESCRIPTION
**Pull request by StanleyGoldman**
Fix for #408 
I do not believe that `Avatar` should be copied in `Account.CopyFrom()`.

### CODE REVIEW

**Comment by grokys** [comment]
Hi @StanleyGoldman.

Could you explain how this fixes the issue? I can see that maybe the problem is the subscription to the `bitmapSource` isn't being copied as well, but simply not copying the avatar doesn't look correct on first glance to me. Surely if the account is different when `CopyTo` is called then the avatar needs to be updated? I think we'd need unit tests here to be sure...

Disclaimer: I did not write this code, so take what I say with a pinch of salt ;)

**Comment by StanleyGoldman** [comment]
Hey @grokys.

After looking at it a bit more, you are definitely right, removing that line from the `CopyFrom` can't be the right fix. I'm trying to better understand why a null is being copied over, that part doesn't make sense.

The `CopyFrom` is called by `TrackingCollection` when `SortedNone`, but in other situations the Avatar is not null. I started reading unit tests for `TrackingCollection` last night, I should be able to reproduce the issue in a unit test when I understand it better.

**Comment by StanleyGoldman** [comment]
Hey @grokys 

I understand the problem now, unfortunately that didn't make the solution pop out at me. Hopefully someone with more project experience can make a recommendation.

The problem is with the `trackingAuthors` and `trackingAssignees` collections in `PullRequestViewModel`. During `Initialize()` as the pullRequests Observable is being consumed Authors and Assignees are being added to those collections with the `AddItem()` method. If a user previously exists in that collection, as is the case in a pull request refresh, `CopyFrom()` is called to update the original Account object.

If `Account.CopyFrom()` is called before the `bitmapSource` observable in the "replacement object" is complete, `CopyFrom()` will copy a null avatar value to the "destination object". By the time `bitmapSource` is complete, the "replacement object" has conceptually been discarded.

I believe I have successfully demonstrated this in the unit test, by intentionally delaying the `bitmapSource` observable in the "replacement object".

What do you think?

**Comment by StanleyGoldman** [comment]
So I was thinking about how to solve this and I had an idea, but it didn't work out as I had hoped.

I remembered that the `TrackingCollection` has a `NewerComparer` which is documented as...

        /// <summary>
        /// Set a comparer that determines whether the item being processed is newer than the same
        /// item seen before. This is to prevent stale items from overriding newer items when data
        /// is coming simultaneously from cache and from live data. Use a timestamp-like comparison
        /// for best results
        /// </summary>

I was hoping that If a `CreatedAt` property were added to `Account` all the instances in a `PullRequest` list would have the same value as they came from the same API response; that in turn would prevent `CopyFrom` from being called.

It turns out that the `TrackingCollection` does not skip elements that have the same value and `CopyFrom` is still called. Is that correct?

Rebased changes to latest master.

File: src/GitHub.Exports.Reactive/Collections/TrackingCollection.cs [hunk 1]
```diff
@@ -540,7 +540,7 @@ ActionData ProcessItem(ActionData data, List<T> list)
                 if (newer != null)
                 {
                     // the object is "older" than the one we have, ignore it
-                    if (newer(item, old) > 0)
+                    if (newer(item, old) >= 0)
```
Comment (StanleyGoldman) [inline src/GitHub.Exports.Reactive/Collections/TrackingCollection.cs]: So this little edge case is the culprit.

**Comment by StanleyGoldman** [comment]
Of course that fix caused a few other unit tests to fail, but I believe there are errors in the tests themselves.

File: src/TrackingCollectionTests/TrackingCollectionTests.cs [hunk 2]
```diff
@@ -333,7 +333,7 @@ public void FilterTitleRun2()
         var total = 1000;
 
         var list1 = new List<Thing>(Enumerable.Range(1, total).Select(i => GetThing(i, i, i, "Run 1")).ToList());
-        var list2 = new List<Thing>(Enumerable.Range(1, total).Select(i => GetThing(i, i, i, "Run 2")).ToList());
+        var list2 = new List<Thing>(Enumerable.Range(1, total).Select(i => GetThing(i, i, i + 1, "Run 2")).ToList());
```
Comment (StanleyGoldman) [inline src/TrackingCollectionTests/TrackingCollectionTests.cs]: So in this unit test, we are expecting the collection to reflect a list or Things with the Title "Run 2".
But if the updated times are the same, that should not be the case, so I am making a change to ensure the "Run 2" things have a newer `UpdatedAt` and now the unit test passes.

File: src/TrackingCollectionTests/TrackingCollectionTests.cs [hunk 3]
```diff
@@ -429,8 +429,8 @@ public void OrderByMatchesOriginalOrder()
         var count = 0;
         var total = 1000;
 
-        var list1 = new List<Thing>(Enumerable.Range(1, total).Select(i => GetThing(i, i, total - i, "Run 1")).ToList());
-        var list2 = new List<Thing>(Enumerable.Range(1, total).Select(i => GetThing(i, i, total - i, "Run 2")).ToList());
+        var list1 = new List<Thing>(Enumerable.Range(1, total).Select(i => GetThing(i, i, i, "Run 1")).ToList());
+        var list2 = new List<Thing>(Enumerable.Range(1, total).Select(i => GetThing(i, i, i + 1, "Run 2")).ToList());
```
Comment (StanleyGoldman) [inline src/TrackingCollectionTests/TrackingCollectionTests.cs]: Again, making sure we have the correct `UpdatedAt` times.

File: src/TrackingCollectionTests/TrackingCollectionTests.cs [hunk 4]
```diff
@@ -464,7 +464,7 @@ public void OrderByMatchesOriginalOrder()
 
         Assert.AreEqual(total, count);
         Assert.AreEqual(total, col.Count);
-        CollectionAssert.AreEqual(col, list2);
+        CollectionAssert.AreEqual(col, list2.Reverse<Thing>());
```
Comment (StanleyGoldman) [inline src/TrackingCollectionTests/TrackingCollectionTests.cs]: This list is being sorted descending so list2 should be reversed.

File: src/GitHub.App/Services/ModelService.cs [hunk 5]
```diff
@@ -164,13 +165,13 @@ public IObservable<AccountCacheItem> GetUserFromCache()
...
                             // this could blow up due to the collection being disposed somewhere else
-                            try { collection.RemoveItem(Create(item)); }
+                            try { collection.RemoveItem(Create(item, DateTime.Now)); }
                             catch (ObjectDisposedException) { }
                         },
                         TimeSpan.Zero,
                         TimeSpan.FromDays(7))
                 )
-                .Select(Create)
+                .Select(prCacheItem => Create(prCacheItem, createdAt))
```
Comment (StanleyGoldman) [inline src/GitHub.App/Services/ModelService.cs]: Setting the create date of a set of pull requests

File: src/GitHub.App/ViewModels/PullRequestListViewModel.cs [hunk 6]
```diff
@@ -68,6 +68,10 @@ public class PullRequestListViewModel : BaseViewModel, IPullRequestListViewModel
                 OrderedComparer<IAccount>.OrderByDescending(x => x.Login).Compare);
             trackingAssignees = new TrackingCollection<IAccount>(Observable.Empty<IAccount>(),
                 OrderedComparer<IAccount>.OrderByDescending(x => x.Login).Compare);
+
+            trackingAuthors.NewerComparer = OrderedComparer<IAccount>.OrderByDescending(x => x.CreatedAt).Compare;
+            trackingAssignees.NewerComparer = OrderedComparer<IAccount>.OrderByDescending(x => x.CreatedAt).Compare;
+
```
Comment (StanleyGoldman) [inline src/GitHub.App/ViewModels/PullRequestListViewModel.cs]: Adding this `NewerComparer` will prevent `CopyFrom()` being called multiple times if a user is the author/assignee of multiple pull requests in the set.

File: src/GitHub.App/Models/Account.cs [hunk 7]
```diff
@@ -75,11 +85,22 @@ public void CopyFrom(IAccount other)
...
+
+            var otherAccount = other as Account;
+            if (otherAccount != null)
+            {
+                _bitmapSourceSubscription.Dispose();
+
+                _bitmapSourceSubscription = otherAccount.BitmapSource
+                    .ObserveOn(RxApp.MainThreadScheduler)
+                    .Subscribe(x => Avatar = x);
+            }
```
Comment (StanleyGoldman) [inline src/GitHub.App/Models/Account.cs]: If we are copying this `Account` from another `Account` instance. We should cancel the subscription to the previous `BitmapSource` and start a new subscription against the new `BitmapSource`.

File: src/GitHub.App/Models/Account.cs [hunk 8]
```diff
@@ -30,9 +35,12 @@ public class Account : ReactiveObject, IAccount
             PrivateReposInPlan = privateRepositoryInPlanCount;
             IsOnFreePlan = privateRepositoryInPlanCount == 0;
             HasMaximumPrivateRepositories = OwnedPrivateRepos >= PrivateReposInPlan;
+            CreatedAt = createdAt ?? DateTime.Now;
+            BitmapSource = bitmapSource;
```
Comment (StanleyGoldman) [inline src/GitHub.App/Models/Account.cs]: Keep track of the BitmapSource in case we are copied to another instance.

File: src/GitHub.App/Models/Account.cs [hunk 9]
```diff
@@ -30,9 +35,12 @@ public class Account : ReactiveObject, IAccount
...
 
-            bitmapSource.ObserveOn(RxApp.MainThreadScheduler)
-                .Subscribe(x => avatar = x);
+            _bitmapSourceSubscription = bitmapSource
+                .ObserveOn(RxApp.MainThreadScheduler)
+                .Subscribe(x => Avatar = x);
```
Comment (StanleyGoldman) [inline src/GitHub.App/Models/Account.cs]: Keep track of the `BitmapSource` `Subscribe()` disposable so we can stop the subscription if we are copied into.

**Comment by StanleyGoldman** [comment]
I'm able to refresh on the pull request screen now.
Refresh

**Comment by StanleyGoldman** [comment]
This also resolves #403

**Comment by grokys** [comment]
Hi Stanley, I've tried to review this PR but I don't quite understand. There seem to be two parts to the fix: the avatar `CopyFrom` fix and changes to `TrackingCollection`'s comparer. I've read your comment above that mentions trying to prevent `CopyFrom` from being called by using a comparer, but I'm not sure how the two parts go together because if you've fixed `CopyFrom` then the `TrackingCollection` changes shouldn't be needed. What am I missing?

**Comment by StanleyGoldman** [comment]
Hey @grokys I think you understand perfectly ;)

You are right, I can definitely (and probably should) separate this pull request into two.

File: src/UnitTests/GitHub.App/Models/AccountModelTests.cs [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+            //A function that will return this image in an observable after X seconds
+            var image = AvatarProvider.CreateBitmapImage("pack://application:,,,/GitHub.App;component/Images/default_user_avatar.png");
+            Func<int, IObservable<BitmapImage>> generateObservable = seconds =>
+            {
+                return Observable.Generate(
+                    initialState: 0, 
+                    condition: i => i < 1, 
+                    iterate: state => state + 1,
+                    resultSelector: i => image, 
+                    timeSelector: i => TimeSpan.FromSeconds(seconds));
```
Comment (StanleyGoldman) [inline src/UnitTests/GitHub.App/Models/AccountModelTests.cs]: Yea, that's probably a better idea. I'll work those changes in.
